File: awdl_connect_oc/utils/NetworkManager.m
//
//  NetworkManager.m
//  awdl_connect_oc
//
//  Created by junjiecui on 2026/4/5.
//

#import "NetworkManager.h"
#import <ifaddrs.h>
#import <arpa/inet.h>
#import <net/if.h>
#import <netdb.h>
#include <sys/socket.h>

// 使用 NSLog 进行日志输出（可替换为 os_log 如果需要）
#define LOG_INFO(fmt, ...) NSLog((fmt), ##__VA_ARGS__)
#define LOG_WARNING(fmt, ...) NSLog((@"⚠️ " fmt), ##__VA_ARGS__)
#define LOG_ERROR(fmt, ...) NSLog((@"❌ " fmt), ##__VA_ARGS__)

@implementation NetworkManager

+ (instancetype)sharedManager {
    static NetworkManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

+ (void)logAWDLIPv6Addresses {
    NSArray<NSString *> *addresses = [self getIPv6AddressesForInterface:@"awdl0"];
    if (addresses.count == 0) {
        LOG_WARNING(@"No IPv6 address found on awdl0 interface");
    } else {
        for (NSString *addr in addresses) {
            LOG_INFO(@"AWDL IPv6 address: %@", addr);
        }
    }
}

+ (NSArray<NSString *> *)getIPv6AddressesForInterface:(NSString *)interfaceName {
    NSMutableArray<NSString *> *addresses = [NSMutableArray array];
    struct ifaddrs *ifaddr = NULL;
    
    if (getifaddrs(&ifaddr) != 0) {
        LOG_ERROR(@"getifaddrs failed");
        return addresses;
    }
    
    for (struct ifaddrs *ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next) {
        NSString *name = [NSString stringWithUTF8String:ifa->ifa_name];
        if (![name isEqualToString:interfaceName]) {
            continue;
        }
        
        struct sockaddr *sa = ifa->ifa_addr;
        if (sa == NULL || sa->sa_family != AF_INET6) {
            continue;
        }
        
        char host[NI_MAXHOST];
        socklen_t salen = (socklen_t)sa->sa_len;
        if (getnameinfo(sa, salen, host, sizeof(host), NULL, 0, NI_NUMERICHOST) == 0) {
            NSString *addrString = [NSString stringWithUTF8String:host];
            [addresses addObject:addrString];
        }
    }
    freeifaddrs(ifaddr);
    return addresses;
}

@end

File: awdl_NW/models/NetworkManager.swift
//
//  NetworkAnalyser.swift
//  awdl_NW
//
//  Created by junjiecui on 2026/4/1.
//
import Network
import os.log

class NetworkManager {
    static let shared = NetworkManager()
    
    private let logger = Logger(subsystem: "com.example.NetworkAnalyser", category: "NetworkAnalyser")
    
    /// Retrieve and log IPv6 addresses of the AWDL interface (awdl0)
    func logAWDLIPv6Addresses() {
        let awdlAddresses = getInterfaceIPv6Addresses(interfaceName: "awdl0")
        if awdlAddresses.isEmpty {
            logger.warning("No IPv6 address found on awdl0 interface")
        } else {
            for addr in awdlAddresses {
                logger.info("AWDL IPv6 address: \(addr)")
            }
        }
    }
    
    /// Enumerate all network interfaces and return IPv6 addresses for the specified interface
    func getInterfaceIPv6Addresses(interfaceName: String) -> [String] {
        var addresses: [String] = []
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        
        guard getifaddrs(&ifaddr) == 0, let firstAddr = ifaddr else {
            logger.error("getifaddrs failed")
            return addresses
        }
        defer { freeifaddrs(ifaddr) }
        
        var current: UnsafeMutablePointer<ifaddrs>? = firstAddr
        while let addr = current {
            let name = String(cString: addr.pointee.ifa_name)
            let family = addr.pointee.ifa_addr.pointee.sa_family
            
            // Filter: match interface name and IPv6 family (AF_INET6 = 30)
            if name == interfaceName && family == UInt8(AF_INET6) {
                var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                let sockAddr = addr.pointee.ifa_addr
                let sockLen = socklen_t(addr.pointee.ifa_addr.pointee.sa_len)
                
                if getnameinfo(sockAddr, sockLen,
                               &hostname, socklen_t(hostname.count),
                               nil, 0, NI_NUMERICHOST) == 0 {
                    let ipv6String = String(cString: hostname)
                    addresses.append(ipv6String)
                }
            }
            current = addr.pointee.ifa_next
        }
        return addresses
    }
    
    /// Resolve IP address by connecting to the discovered service endpoint
    func resolveEndpoint(_ endpoint: NWEndpoint) {
        let connection = NWConnection(to: endpoint, using: .tcp)
        connection.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                // Once connected, we can get the resolved remote endpoint with IP
                if let remotePath = connection.currentPath,
                   let remoteEndpoint = remotePath.remoteEndpoint {
                    self?.logger.info("Resolved IP: \(remoteEndpoint.debugDescription)")
                }
                // Get local address info as well
                if let localEndpoint = connection.currentPath?.localEndpoint {
                    self?.logger.info("Local endpoint: \(localEndpoint.debugDescription)")
                }
                connection.cancel()
            case .failed(let error):
                self?.logger.error("Connection failed: \(error)")
                connection.cancel()
            default:
                break
            }
        }
        connection.start(queue: .main)
    }
}
